File: rust/src/trie.rs
use crate::trie_node::TrieNode;
pub struct Trie<TValue: Clone> {
    nodes: Vec<TrieNode<TValue>>,
}

impl<TValue: Clone> Trie<TValue> {
    /// Initializes a new, empty Trie.
    pub fn new() -> Self {
        let mut nodes = Vec::new();
        nodes.push(TrieNode::new());
        Self { nodes }
    }

    pub fn with_capacity(cap: usize) -> Self {
        let mut nodes = Vec::with_capacity(cap);
        nodes.push(TrieNode::new());
        Self { nodes }
    }

    fn alloc_node(&mut self) -> u32 {
        let idx = self.nodes.len() as u32;
        self.nodes.push(TrieNode::new());
        idx
    }

    /// Inserts a key-value pair into the Trie.
    /// value is used to mark the end of the string or can contain a value if using the trie as a /// key value pair.
    ///  If the key already exists,
    /// its value is updated.
    ///
    /// Example:
    /// ```Rust
    /// let mut trie = Trie::new();
    /// trie.insert("apple", 1);
    /// assert_eq!(trie.get("apple"), Some(&1));
    /// ```
    pub fn insert(&mut self, key: &str, value: &TValue) {
        let mut current = 0u32; // root
        for c in key.chars() {
            let next = match self.nodes[current as usize].get_child(c) {
                Some(idx) => idx,
                None => {
                    let idx = self.alloc_node();
                    self.nodes[current as usize].add_child(c, idx);
                    idx
                }
            };
            current = next;
        }
        self.nodes[current as usize].set_value(value.clone());
    }

    /// Searches for a key and returns a reference to its value if it exists.
    ///
    /// Example:
    /// ```Rust
    /// let mut trie = Trie::new();
    /// trie.insert("apple", 1);
    /// assert_eq!(trie.get("apple"), Some(&1));
    /// ```
    pub fn get(&self, key: &str) -> Option<&TValue> {
        let mut current = 0u32;
        for c in key.chars() {
            match self.nodes[current as usize].get_child(c) {
                Some(idx) => current = idx,
                None => return None,
            }
        }
        self.nodes[current as usize].get_value()
    }

    /// Deletes a key and its associated value from the Trie.
    /// Returns true if the key was found and deleted, false otherwise.
    ///
    /// Example:
    /// ```Rust
    /// let mut trie = Trie::new();
    /// trie.insert("apple", 1);
    /// assert_eq!(trie.delete("apple"), true);
    /// assert_eq!(trie.get("apple"), None);
    /// ```
    pub fn delete(&mut self, key: &str) -> bool {
        if key.is_empty() {
            return false;
        }

        // Walk down, recording the path as (node_index, char) pairs
        let mut path: Vec<(u32, char)> = Vec::new();
        let mut current = 0u32;

        for c in key.chars() {
            match self.nodes[current as usize].get_child(c) {
                Some(idx) => {
                    path.push((current, c));
                    current = idx;
                }
                None => return false, // key doesn't exist
            }
        }

        // current is now the terminal node — check it's actually a word
        if !self.nodes[current as usize].is_end_of_word() {
            return false;
        }

        self.nodes[current as usize].clear_value();

        // Walk back up pruning nodes that are now dead
        // (no value, no children)
        let mut child = current;
        for (parent, c) in path.into_iter().rev() {
            if self.nodes[child as usize].has_children()
                || self.nodes[child as usize].is_end_of_word()
            {
                break; // still needed, stop pruning
            }
            self.nodes[parent as usize].remove_child(c);
            child = parent;
        }

        true
    }
    /// Checks if there is any word in the trie that starts with the given prefix.
    /// Returns true if such a prefix exists, false otherwise.
    /// Example:
    /// ```Rust
    /// let mut trie = Trie::new();
    /// trie.insert("apple", 1);
    /// assert_eq!(trie.prefix_search("app"), true);
    /// assert_eq!(trie.prefix_search("apl"), false);
    /// ```
    pub fn prefix_search(&self, prefix: &str) -> bool {
        let mut current = 0u32;
        for c in prefix.chars() {
            match self.nodes[current as usize].get_child(c) {
                Some(idx) => current = idx,
                None => return false,
            }
        }
        true
    }

    /// Returns up to `max_results` words in the trie that start with the given prefix.
    /// If no words match the prefix, returns an empty vector.
    /// Example:
    /// ```Rust
    /// let mut trie = Trie::new();
    /// trie.insert("apple", 1);
    /// trie.insert("app", 2);
    /// trie.insert("apricot", 3);
    /// let results = trie.auto_complete("ap", 2);
    /// assert_eq!(results, vec!["app", "apple"]);
    /// ```
    pub fn auto_complete(&self, prefix: &str, max_results: usize) -> Vec<String> {
        let mut results = Vec::new();
        if max_results == 0 {
            return results;
        }

        let mut current = 0u32;
        for c in prefix.chars() {
            match self.nodes[current as usize].get_child(c) {
                Some(idx) => current = idx,
                None => return results,
            }
        }

        if self.nodes[current as usize].is_end_of_word() {
            results.push(prefix.to_string());
            if results.len() >= max_results {
                return results;
            }
        }

        Self::collect_words_recursive(
            &self.nodes,
            current,
            prefix.to_string(),
            &mut results,
            max_results,
        );
        results
    }

    fn collect_words_recursive(
        nodes: &[TrieNode<TValue>],
        current: u32,
        curr_prefix: String,
        results: &mut Vec<String>,
        max_results: usize,
    ) {
        if results.len() >= max_results {
            return;
        }
        for (c, idx) in nodes[current as usize].children_iter() {
            if results.len() >= max_results {
                return;
            }
            let new_prefix = format!("{curr_prefix}{c}");
            if nodes[idx as usize].is_end_of_word() {
                results.push(new_prefix.clone());
            }
            Self::collect_words_recursive(nodes, idx, new_prefix, results, max_results);
        }
    }

    /// Adds multiple words to the trie from a list, using a value generator function
    /// to determine the value associated with each word.
    /// This is useful for bulk insertion where the value might depend on the word itself.
    /// Example:
    /// ```Rust
    /// let mut trie = Trie::new();
    /// trie.add_word_list(&["cat", "car", "cart"], |word| word.len());
    /// assert_eq!(trie.get("cat"), Some(&3));
    /// assert_eq!(trie.get("car"), Some(&3));
    /// assert_eq!(trie.get("cart"), Some(&4));
    /// ```
    pub fn add_word_list<T, F>(&mut self, items: &[T], value_generator: F)
    where
        T: AsRef<str>,
        F: Fn(&T) -> TValue,
    {
        for item in items {
            self.insert(item.as_ref(), &value_generator(item));
        }
    }
}

/// Allows creating a new Trie with `Trie::default()`.
impl<TValue: Clone> Default for Trie<TValue> {
    fn default() -> Self {
        Self::new()
    }
}

File: go/trie.go
package trie

import (
	"errors"
)

type TrieError int

var TrieErrorChildDoesNotExist = errors.New("No child available")

type Trie[TValue any] struct {
	// The root node of the Trie. It does not hold any character itself.
	root *TrieNode[TValue]
}

// Insert adds a key-value pair into the Trie.
// The value marks the end of a string and can be retrieved later.
// If the key already exists, its value is updated.
//
// Example:
//
//	trie := NewTrie()
//	trie.Insert("apple", 1)
//	value, exists := trie.Get("apple")
//	// value == 1, exists == true
func (self *Trie[TValue]) Insert(key *string, value TValue) {
	currentNode := self.root
	for _, char := range *key {
		if !currentNode.containsChild(char) {
			currentNode.addChild(char)
		}
		child, _ := currentNode.getChildMut(char)
		currentNode = child
	}
	currentNode.setValue(value)
}

// Get searches for a key and returns its value if it exists.
// It returns the value and a boolean indicating whether the key was found.
//
// Example:
//
//	trie := NewTrie()
//	trie.Insert("apple", 1)
//	value, exists := trie.Get("apple") // value == 1, exists == true
func (self *Trie[TValue]) Get(key *string) (*TValue, bool) {
	currentNode := self.root
	for _, char := range *key {
		if child, err := currentNode.getChildMut(char); err == nil {
			currentNode = child
		} else {
			return nil, false
		}
	}

	return currentNode.getValue()
}

// Delete removes a key and its associated value from the Trie.
// It returns true if the key was found and deleted, false otherwise.
//
// Example:
//
//	trie := NewTrie()
//	trie.Insert("apple", 1)
//	deleted := trie.Delete("apple") // deleted == true
//	value, exists := trie.Get("apple") // exists == false
func (self *Trie[TValue]) Delete(key *string) bool {
	if key == nil || *key == "" {
		// Empty key is not allowed
		return false
	}
	deleted, _ := self.deleteRecursive(self.root, *key)
	return deleted
}

func (self *Trie[TValue]) deleteRecursive(currentNode *TrieNode[TValue], key string) (bool, bool) {
	if key == "" {

		// We have reached the node corresponding to the key.
		if currentNode.isEndOfWord() {
			currentNode.clearValue()

			// Return true if this node has no children, so the parent can remove it.
			return true, !currentNode.hasChildren()
		}
		// Key doesn't actually exist as a word in the trie.
		return false, false
	}

	runes := []rune(key)
	c := runes[0]
	remaining := string(runes[1:])
	if child, err := currentNode.getChildMut(c); err == nil {
		deleted, shouldDeleteChild := self.deleteRecursive(child, remaining)
		if shouldDeleteChild {
			currentNode.removeChild(c)
		}

		shouldPruneThisNode := deleted && !currentNode.isEndOfWord() && !currentNode.hasChildren()
		return deleted, shouldPruneThisNode
	} else {
		// The path for the key does not exist.
		return false, false
	}
}

func (self *Trie[TValue]) PrefixSearch(prefix *string) bool {
	currentNode := self.root // Use local variable!
	for _, char := range *prefix {
		if child, err := currentNode.getChildMut(char); err == nil {
			currentNode = child // Modify local variable, not self.root
		} else {
			return false
		}
	}
	return true // If we got here, prefix exists
}

// Preforms a prefix serach on the trie and returns all words that start with the given prefix.
// If no words match the prefix, an empty list is returned.
//
// Example:
//
//	trie := NewTrie()
//	trie.Insert("apple", 1)
//	trie.Insert("app", 2)
//	words := trie.AutoComplete("app") // words == ["app", "apple"]
func (self *Trie[TValue]) AutoComplete(prefix *string) []string {
	results := []string{}
	currentNode := self.root
	for _, char := range *prefix {
		if child, err := currentNode.getChildMut(char); err == nil {
			currentNode = child
		} else {
			// Prefix not found, return empty list
			return results
		}
	}

	self.collectWordsRecursive(currentNode, *prefix, &results)
	return results
}

// Helper function to recursively collect words from the trie.
// Adds prefix to results if prefix is  a complete word, then continues to explore children to find
// extensions of the prefix.
func (self *Trie[TValue]) collectWordsRecursive(node *TrieNode[TValue], currPrefix string, results *[]string) {
	// If the current node marks the end of a word, add the current prefix to results.
	if node.isEndOfWord() {
		*results = append(*results, currPrefix)
	}
	// Continue with all children to find extensions.
	for char, child := range node.children {
		newPrefix := currPrefix + string(char)
		self.collectWordsRecursive(child, newPrefix, results)
	}
}

// Takes a list of words and a function that generates a value for each word to store at the end of
// the chain.
func (self *Trie[TValue]) AddWordList(words *[]string, valueGenerator func(string) TValue) {
	for _, word := range *words {
		self.Insert(&word, valueGenerator(word))
	}
}
